micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java

```diff
@@ -94,12 +94,21 @@ public abstract class MeterRegistry {
 
     private final More more = new More();
 
-    // Even though writes are guarded by meterMapLock, iterators across value space are
-    // supported
-    // Hence, we use CHM to support that iteration without ConcurrentModificationException
-    // risk
+    /**
+     * Even though writes are guarded by meterMapLock, iterators across value space are
+     * supported. Hence, we use CHM to support that iteration without
+     * ConcurrentModificationException risk.
+     */
     private final Map<Id, Meter> meterMap = new ConcurrentHashMap<>();
 
+    // writes guarded by meterMapLock, reads can be stale
+    // VisibleForTesting
+    final Map<Id, Meter> preFilterIdToMeterMap = new HashMap<>();
+
+    // not thread safe; only needed when MeterFilter configured with Meters registered
+    // VisibleForTesting
+    final Set<Id> stalePreFilterIds = new HashSet<>();
```

Comment:
Also, should we make setters that return unmodifiable collections and make these collections private?

Reply:
I assume you meant getters. That's a good idea. I've done it in https://github.com/micrometer-metrics/micrometer/commit/f1471c573d2067ca508387f67815b6b5858d53fc